tests/datacontroller/insertcontent.js

```diff
@@ -0,0 +1,606 @@
+				test(
+					'inserts text',
+					'xxx',
+					'<paragraph>foo[<inlineWidget></inlineWidget>]bar</paragraph>',
+					'<paragraph>fooxxx[]bar</paragraph>'
+				);
+
+				test(
+					'inserts paragraph',
+					'<paragraph>xxx</paragraph>',
+					'<paragraph>foo[<inlineWidget></inlineWidget>]bar</paragraph>',
+					'<paragraph>fooxxx[]bar</paragraph>'
+				);
+
+				test(
+					'inserts text + paragraph',
+					'yyy<paragraph>xxx</paragraph>',
+					'<paragraph>foo[<inlineWidget></inlineWidget>]bar</paragraph>',
+					'<paragraph>fooyyy</paragraph><paragraph>xxx[]bar</paragraph>'
+				);
+
+				test(
+					'inserts two blocks',
+					'<heading1>xxx</heading1><paragraph>yyy</paragraph>',
+					'<paragraph>foo[<inlineWidget></inlineWidget>]bar</paragraph>',
+					'<paragraph>fooxxx</paragraph><paragraph>yyy[]bar</paragraph>'
+				);
+
+				test(
+					'inserts inline object',
+					'<inlineWidget></inlineWidget>',
+					'<paragraph>foo[<inlineWidget></inlineWidget>]bar</paragraph>',
+					'<paragraph>foo<inlineWidget></inlineWidget>[]bar</paragraph>'
+				);
+
+				test(
+					'inserts block object',
+					'<blockWidget></blockWidget>',
+					'<paragraph>foo[<inlineWidget></inlineWidget>]bar</paragraph>',
+					'<paragraph>foo</paragraph>[<blockWidget></blockWidget>]<paragraph>bar</paragraph>'
+				);
+			} );
+		} );
+
+		describe( 'filtering out', () => {
+			beforeEach( () => {
+				doc = new Document();
+				doc.createRoot();
+
+				dataController = new DataController( doc );
+
+				const schema = doc.schema;
+
+				schema.registerItem( 'paragraph', '$block' );
+
+				// Let's use table as an example of content which needs to be filtered out.
+				schema.registerItem( 'table' );
+				schema.registerItem( 'td' );
+				schema.registerItem( 'disallowedWidget' );
+
+				schema.allow( { name: 'table', inside: '$clipboardHolder' } );
+				schema.allow( { name: 'td', inside: '$clipboardHolder' } );
+				schema.allow( { name: '$block', inside: 'td' } );
+				schema.allow( { name: '$text', inside: 'td' } );
+
+				schema.allow( { name: 'disallowedWidget', inside: '$clipboardHolder' } );
+				schema.allow( { name: '$text', inside: 'disallowedWidget' } );
+				schema.objects.add( 'disallowedWidget' );
+			} );
+
+			test(
+				'filters out disallowed elements and leaves out the text',
+				'<table><td>xxx</td><td>yyy</td></table>',
+				'<paragraph>f[]oo</paragraph>',
+				'<paragraph>fxxxyyy[]oo</paragraph>'
+			);
+
+			test(
+				'filters out disallowed elements and leaves out the paragraphs',
+				'<table><td><paragraph>xxx</paragraph><paragraph>yyy</paragraph><paragraph>zzz</paragraph></td></table>',
+				'<paragraph>f[]oo</paragraph>',
+				'<paragraph>fxxx</paragraph><paragraph>yyy</paragraph><paragraph>zzz[]oo</paragraph>'
+			);
+
+			test(
+				'filters out disallowed objects',
+				'<disallowedWidget>xxx</disallowedWidget>',
+				'<paragraph>f[]oo</paragraph>',
+				'<paragraph>f[]oo</paragraph>'
+			);
```

Comment:
You know that "helpers better then magic" is true also in tests? Such tests are very hard  to maintain. You already had to create a `options.beforeCallback` method which is not very readable. In fact, these tests are not readable at all. And I think that it's worth to make code even longer if it is more readable thanks to it, but in this case, code is even shorter:

```js
it( 'filters out disallowed objects', () => {
	setData( doc, '<disallowedWidget>xxx</disallowedWidget>' );
	insertHelper( '<paragraph>f[]oo</paragraph>' );
	expect( getData( doc ) ).to.equal( '<paragraph>f[]oo</paragraph>' );
} );

function insertHelper( content ) {
		if ( typeof content == 'string' ) {
			content = parse( content, doc.schema, {
				context: [ '$clipboardHolder' ]
			} );
		}

		if ( !( content instanceof ModelDocumentFragment ) ) {
			content = new ModelDocumentFragment( [ content ] );
		}

		// Override the convertion so we get exactly the model that we defined in the content param.
		// This way we avoid the need to write converters for everything we want to test.
		dataController.viewToModel.convert = () => {
			return content;
		};

		insertContent( dataController, doc.batch(), doc.selection, new ViewDocumentFragment() );
}
```